# 3. Import your analysis libraries, as well as your new customer data set as a dataframe.

## 01. Import libraries and data

In [1]:
# Import libraries

import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

In [2]:
# Turn folder path into a string 

path = r'C:\Users\carol\Documents\DA\Data Inmersion\04 Python Fundamentals for Data Analysts\Instacart Basket Analysis'

In [3]:
# Import customers.csv

df_customers = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'customers.csv'), index_col = False)

In [4]:
df_customers.shape

(206209, 10)

# 4. Wrangle the data so that it follows consistent logic

In [5]:
# Check dataframe

df_customers.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

In [6]:
# Change name of column "First Name" to first_name

df_customers.rename(columns = {'First Name' : 'first_name'}, inplace = True)

In [7]:
# Change name of column "Surnam" to Last Name

df_customers.rename(columns = {'Surnam' : 'last_name'}, inplace = True)

In [8]:
# Change name of column "Gender" to Last Name

df_customers.rename(columns = {'Gender' : 'gender'}, inplace = True)

In [9]:
# Change name of column "STATE" to State

df_customers.rename(columns = {'STATE' : 'state'}, inplace = True)

In [10]:
# Change name of column "Age" to age

df_customers.rename(columns = {'Age' : 'age'}, inplace = True)

In [11]:
# Change name of column "_dependants" to number_of_dependants

df_customers.rename(columns = {'n_dependants' : 'number_of_dependants'}, inplace = True)

In [12]:
# Change name of column "_dependants" to marital_status

df_customers.rename(columns = {'fam_status' : 'marital_status'}, inplace = True)

In [13]:
#Check output

df_customers.head()

user_id first_name last_name  gender       state  age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   number_of_dependants marital_status  income  
0                     3        married  165665  
1                     0         single   59285  
2                     2        married   99568  
3                     0         single   42049  
4                     1        married   40374

In [14]:
# Check data types

df_customers.dtypes

user_id                  int64
first_name              object
last_name               object
gender                  object
state                   object
age                      int64
date_joined             object
number_of_dependants     int64
marital_status          object
income                   int64
dtype: object

In [15]:
# change the data type of the "user_id" column from a number to a string

df_customers['user_id'] = df_customers['user_id'].astype('str')

In [16]:
# Check output

df_customers['user_id'].dtype

dtype('O')

# 5. Complete the fundamental data quality and consistency checks 

In [17]:
# Check general stats

df_customers.describe()

age  number_of_dependants         income
count  206209.000000         206209.000000  206209.000000
mean       49.501646              1.499823   94632.852548
std        18.480962              1.118433   42473.786988
min        18.000000              0.000000   25903.000000
25%        33.000000              0.000000   59874.000000
50%        49.000000              1.000000   93547.000000
75%        66.000000              3.000000  124244.000000
max        81.000000              3.000000  593901.000000

### There's nothing remarkable in the stats. Age is between 18-81, number of dependants is between 0-3, and income is between 25.9K - 593K.

In [18]:
# Find missing values

df_customers.isnull().sum()

user_id                     0
first_name              11259
last_name                   0
gender                      0
state                       0
age                         0
date_joined                 0
number_of_dependants        0
marital_status              0
income                      0
dtype: int64

### There are 11,259 first names missing. The analysis is not impacted by this so no action will be taken.

In [19]:
# Check for mixed types

for col in df_customers.columns.tolist():
  weird = (df_customers[[col]].applymap(type) != df_customers[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_customers[weird]) > 0:
    print (col)

first_name


In [20]:
# Convert to string

df_customers['first_name'] = df_customers['first_name'].astype('str')

In [21]:
# Check again for mixed types

for col in df_customers.columns.tolist():
  weird = (df_customers[[col]].applymap(type) != df_customers[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_customers[weird]) > 0:
    print (col)

In [22]:
# look for full duplicates within the dataframe

df_dups = df_customers[df_customers.duplicated()]

In [23]:
df_dups

Empty DataFrame
Columns: [user_id, first_name, last_name, gender, state, age, date_joined, number_of_dependants, marital_status, income]
Index: []

### There are no duplicates

In [24]:
df_customers.columns

Index(['user_id', 'first_name', 'last_name', 'gender', 'state', 'age',
       'date_joined', 'number_of_dependants', 'marital_status', 'income'],
      dtype='object')

In [25]:
df_customers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206209 entries, 0 to 206208
Data columns (total 10 columns):
 #   Column                Non-Null Count   Dtype 
---  ------                --------------   ----- 
 0   user_id               206209 non-null  object
 1   first_name            206209 non-null  object
 2   last_name             206209 non-null  object
 3   gender                206209 non-null  object
 4   state                 206209 non-null  object
 5   age                   206209 non-null  int64 
 6   date_joined           206209 non-null  object
 7   number_of_dependants  206209 non-null  int64 
 8   marital_status        206209 non-null  object
 9   income                206209 non-null  int64 
dtypes: int64(3), object(7)
memory usage: 15.7+ MB


In [26]:
# Change types for ords data set

df_customers['age']=df_customers['age'].astype('int8')
df_customers['number_of_dependants']=df_customers['number_of_dependants'].astype('int8')

In [27]:
df_customers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206209 entries, 0 to 206208
Data columns (total 10 columns):
 #   Column                Non-Null Count   Dtype 
---  ------                --------------   ----- 
 0   user_id               206209 non-null  object
 1   first_name            206209 non-null  object
 2   last_name             206209 non-null  object
 3   gender                206209 non-null  object
 4   state                 206209 non-null  object
 5   age                   206209 non-null  int8  
 6   date_joined           206209 non-null  object
 7   number_of_dependants  206209 non-null  int8  
 8   marital_status        206209 non-null  object
 9   income                206209 non-null  int64 
dtypes: int64(1), int8(2), object(7)
memory usage: 13.0+ MB


In [28]:
df_customers.head()

user_id first_name last_name  gender       state  age date_joined  \
0   26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1   33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2   65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3  125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4  130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   number_of_dependants marital_status  income  
0                     3        married  165665  
1                     0         single   59285  
2                     2        married   99568  
3                     0         single   42049  
4                     1        married   40374

# 6. Combine your customer data with the rest of your prepared Instacart data.

In [29]:
# Import ords_prods_merged.pkl

ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_3.pkl'))

In [30]:
ords_prods_merge['prices'].describe()

count    3.240486e+07
mean     1.198023e+01
std      4.956554e+02
min      1.000000e+00
25%      4.200000e+00
50%      7.400000e+00
75%      1.130000e+01
max      9.999900e+04
Name: prices, dtype: float64

In [31]:
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_last_order  product_id  add_to_cart_order  reordered  \
0                    NaN         196                  1          0   
1                   15.0         196                  1          1   
2                   21.0         196                  1          1   
3                   29.0         196                  1          1   
4                   28.0         196                  1          1   

  product_name  aisle_id  department_id  prices    price_range_loc  \
0         Soda        77              7     9.0  Mid-range product   
1         Soda        77              7     9.0  Mid-range product   
2         Soda        77              7     9.0  Mid-range product   
3         Soda        77              7     9.0  Mid-range product   
4         Soda        77              7     9.0  Mid-range product   

     busiest_days busiest_period_of_day  loyalty_flag spending_flag  \
0  Regularly busy        Average orders  New customer   Low spender   
1    Slowest days        Average orders  New customer   Low spender   
2    Slowest days           Most orders  New customer   Low spender   
3    Slowest days        Average orders  New customer   Low spender   
4    Slowest days           Most orders  New customer   Low spender   

          frequency_flag  
0  Non-frequent customer  
1  Non-frequent customer  
2  Non-frequent customer  
3  Non-frequent customer  
4  Non-frequent customer

In [32]:
# Check output

ords_prods_merge['user_id'].dtype

dtype('int32')

In [33]:
# change the data type of the "user_id" column from a number to a string

ords_prods_merge['user_id'] = ords_prods_merge['user_id'].astype('str')

In [34]:
# Check output

ords_prods_merge['user_id'].dtype

dtype('O')

In [35]:
ords_prods_merge.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 32404859 entries, 0 to 32404858
Data columns (total 19 columns):
 #   Column                 Dtype  
---  ------                 -----  
 0   order_id               int32  
 1   user_id                object 
 2   order_number           int8   
 3   orders_day_of_week     int8   
 4   order_hour_of_day      int8   
 5   days_since_last_order  float16
 6   product_id             int32  
 7   add_to_cart_order      int32  
 8   reordered              int8   
 9   product_name           object 
 10  aisle_id               int64  
 11  department_id          int64  
 12  prices                 float64
 13  price_range_loc        object 
 14  busiest_days           object 
 15  busiest_period_of_day  object 
 16  loyalty_flag           object 
 17  spending_flag          object 
 18  frequency_flag         object 
dtypes: float16(1), float64(1), int32(3), int64(2), int8(4), object(8)
memory usage: 3.4+ GB


In [36]:
# Change types for data set

ords_prods_merge['add_to_cart_order']=ords_prods_merge['add_to_cart_order'].astype('int8')
ords_prods_merge['aisle_id']=ords_prods_merge['aisle_id'].astype('int8')
ords_prods_merge['department_id']=ords_prods_merge['department_id'].astype('int8')
ords_prods_merge['prices']=ords_prods_merge['prices'].astype('float32')

In [37]:
ords_prods_merge['prices'].describe()

count    3.240486e+07
mean     1.198023e+01
std      4.956394e+02
min      1.000000e+00
25%      4.200000e+00
50%      7.400000e+00
75%      1.130000e+01
max      9.999900e+04
Name: prices, dtype: float64

In [38]:
ords_prods_merge.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 32404859 entries, 0 to 32404858
Data columns (total 19 columns):
 #   Column                 Dtype  
---  ------                 -----  
 0   order_id               int32  
 1   user_id                object 
 2   order_number           int8   
 3   orders_day_of_week     int8   
 4   order_hour_of_day      int8   
 5   days_since_last_order  float16
 6   product_id             int32  
 7   add_to_cart_order      int8   
 8   reordered              int8   
 9   product_name           object 
 10  aisle_id               int8   
 11  department_id          int8   
 12  prices                 float32
 13  price_range_loc        object 
 14  busiest_days           object 
 15  busiest_period_of_day  object 
 16  loyalty_flag           object 
 17  spending_flag          object 
 18  frequency_flag         object 
dtypes: float16(1), float32(1), int32(2), int8(7), object(8)
memory usage: 2.8+ GB


In [39]:
# Turn values >100 NaNs

ords_prods_merge.loc[ords_prods_merge['prices'] >100, 'prices'] = np.nan

In [40]:
ords_prods_merge['prices'].describe()

count    3.239973e+07
mean     7.790978e+00
std      4.120040e+00
min      1.000000e+00
25%      4.200000e+00
50%      7.400000e+00
75%      1.130000e+01
max      2.500000e+01
Name: prices, dtype: float64

In [41]:
# Merge

ords_prods_customs_merge = ords_prods_merge.merge(df_customers, on = 'user_id')

In [42]:
# Check output

ords_prods_customs_merge.head()

order_id user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329       1             1                   2                  8   
1   2398795       1             2                   3                  7   
2    473747       1             3                   3                 12   
3   2254736       1             4                   4                  7   
4    431534       1             5                   4                 15   

   days_since_last_order  product_id  add_to_cart_order  reordered  \
0                    NaN         196                  1          0   
1                   15.0         196                  1          1   
2                   21.0         196                  1          1   
3                   29.0         196                  1          1   
4                   28.0         196                  1          1   

  product_name  ...         frequency_flag  first_name  last_name  gender  \
0         Soda  ...  Non-frequent customer       Linda     Nguyen  Female   
1         Soda  ...  Non-frequent customer       Linda     Nguyen  Female   
2         Soda  ...  Non-frequent customer       Linda     Nguyen  Female   
3         Soda  ...  Non-frequent customer       Linda     Nguyen  Female   
4         Soda  ...  Non-frequent customer       Linda     Nguyen  Female   

     state age date_joined number_of_dependants marital_status income  
0  Alabama  31   2/17/2019                    3        married  40423  
1  Alabama  31   2/17/2019                    3        married  40423  
2  Alabama  31   2/17/2019                    3        married  40423  
3  Alabama  31   2/17/2019                    3        married  40423  
4  Alabama  31   2/17/2019                    3        married  40423  

[5 rows x 28 columns]

In [43]:
# Export combined customer data set with ords_prods_group to pickle file

ords_prods_customs_merge.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_customs_merged.pkl'))